Modules/_threadmodule.c

```diff
@@ -1252,6 +1307,89 @@ PyDoc_STRVAR(exit_doc,
 This is synonymous to ``raise SystemExit''.  It will cause the current\n\
 thread to exit silently unless the exception is caught.");
 
+static PyObject *
+thread_PyThread_join_thread(PyObject *module, PyObject *ident_obj)
+{
+    thread_module_state *state = get_thread_state(module);
+
+    if (!PyLong_Check(ident_obj)) {
+        PyErr_Format(PyExc_TypeError, "thread ident must be an int, not %s",
+                     Py_TYPE(ident_obj)->tp_name);
+        return NULL;
+    }
+    // Check if the ident is part of the joinable threads and fetch its handle.
+    unsigned long ident;
+    Py_uintptr_t handle;
+    if (get_joinable_thread(state, ident_obj, &ident, &handle)) {
+        return NULL;
+    }
+    if (ident == PyThread_get_thread_ident()) {
+        // PyThread_join_thread() would deadlock or error out.
+        PyErr_SetString(ThreadError, "Cannot join current thread");
+        return NULL;
+    }
+    // Before actually joining, we must first remove the ident from joinable_dict,
+    // as joining several times simultaneously or sequentially is undefined behavior.
+    if (PyDict_DelItem(state->joinable_dict, ident_obj)) {
+        return NULL;
+    }
+    int ret;
+    Py_BEGIN_ALLOW_THREADS
+    ret = PyThread_join_thread((Py_uintptr_t) handle);
+    Py_END_ALLOW_THREADS
+    if (ret) {
+        PyErr_SetString(ThreadError, "Failed joining thread");
+        return NULL;
+    }
+    Py_RETURN_NONE;
+}
+
+PyDoc_STRVAR(join_thread_doc,
+"join_thread(ident)\n\
+\n\
+Join the thread with the given identifier. The thread must have been started\n\
+with start_new_thread() with the `joinable` argument set to True.\n\
+This function can only be called once per joinable thread, and it cannot be\n\
+called if detach_thread() was previously called on the same thread.\n");
+
+static PyObject *
+thread_PyThread_detach_thread(PyObject *module, PyObject *ident_obj)
+{
+    thread_module_state *state = get_thread_state(module);
+
+    if (!PyLong_Check(ident_obj)) {
+        PyErr_Format(PyExc_TypeError, "thread ident must be an int, not %s",
+                     Py_TYPE(ident_obj)->tp_name);
+        return NULL;
+    }
+    unsigned long ident;
+    Py_uintptr_t handle;
+    if (get_joinable_thread(state, ident_obj, &ident, &handle)) {
+        return NULL;
+    }
+    if (PyDict_DelItem(state->joinable_dict, ident_obj)) {
+        if (PyErr_ExceptionMatches(PyExc_KeyError)) {
+            PyErr_SetString(PyExc_ValueError,
+                            "the given thread is not joinable and thus cannot be detached");
+        }
+        return NULL;
+    }
+    int ret = PyThread_detach_thread(handle);
```

Comment:
(Just to demonstrate my earlier comment about making the thread ID a struct...)

Perhaps we wouldn't need `state->joinable_dict` if the handle were a struct with a `joinable` field and a `detached` field:

```suggestion
    if (PyLong_Check(ident_obj)) {
        // Do some fallback conversion here.
    }
    else if (Py_TYPE(ident_obj) != _PyThreadIDType) {
        PyErr_Format(PyExc_TypeError, "thread ident must be a ThreadID object, not %s",
                     Py_TYPE(ident_obj)->tp_name);
        return NULL;
    }
    _PyThread_handle_t *th = ((_PyThreadIDObject *)ident_obj)->handle;
    if (!th->joinable) {
        PyErr_SetString(PyExc_ValueError, "not a joinable thread");
        return NULL;
    }
    if (th->detached) {
        PyErr_SetString(PyExc_ValueError, "thread already detached");
        return NULL;
    }
    int ret = PyThread_detach_thread(th->handle);
```

Of course, that would require adding `_PyThreadIDType` (and making it look like an `int` object?) and percolating it through the code.  The value would be in any improved conceptual consistency and maintainability.

Reply:
The same idea would apply in `thread_PyThread_join_thread()`.

Reply:
Hmm, this would make it a PyObject, right? That doesn't sound desirable in the PyThread APIs, if I'm understanding your proposal correctly.

Also, again IIUC, in your proposal `start_new_thread` would start returning those integer-like structs instead of plain platform-specific thread ids? This could break user code.

Reply:
That said, we could:
1) keep the low-level PyThread APIs as-is
2) create a new `start_joinable_thread` function instead of adding an optional `joinable` argument to `start_new_thread`
3) have `start_joinable_thread` return a rich object with API methods to detach or join, a finalizer that detaches the thread, and a property to get the thread ident as an integer

That definitely sounds more elegant and robust than the current solution. Hopefully it will also work correctly :-)

Reply:
> Hmm, this would make it a PyObject, right? That doesn't sound desirable in the PyThread APIs, if I'm understanding your proposal correctly.

The PyObject for it would be separate.  For the PyThread APIs we would have a plain struct (though it would only be used for new/internal API, for now at least):

```c
typedef struct pythread_handle {
    HANDLE handle;  // Windows
    pthread_t handle; // other

    unsigned long legacy_ident;

    unsigned int joinable:1;
    unsigned int attached:1;  // neither detached nor joined
} PyThread_handle_t;

int PyThread_start_joinable_thread(void (*func)(void *), void *arg, _PyThread_handle_t* handle);
int PyThread_join_thread(_PyThread_handle_t* handle);
int PyThread_detach_thread(_PyThread_handle_t* handle);
```

`PyThread_start_joinable_thread()` would set all the fields, including setting `handle->joinable = 1;` and `handle->attached = 1;`.

If we were to make this public API, we'd make the struct opaque there (to avoid our sad history of exposing ABI we later wish we hadn't).

> Also, again IIUC, in your proposal `start_new_thread` would start returning those integer-like structs instead of plain platform-specific thread ids? This could break user code.

Yeah, `start_new_thread()` should return the new compound object, but keeping it backward compatible shouldn't be a problem. 🤞

We would have something like `PyThreadIDObject`:

```c
typedef struct threadid {
    PyObject_HEAD
    PyThread_handle_t handle;
} PyThreadIDObject;
```

In Modules/_threadmodule.c, we'd start the thread like this (from [another comment](https://github.com/python/cpython/pull/110848/files/e2f440f893aaa3118fa1bac59914e6c88eed738e#r1373796734)):

```C
    _PyThreadIDObject *ident_obj = _PyThreadID_New();
    if (ident_obj == NULL) {
        goto error;
    }
    int res;
    if (joinable == Py_True) {
        res = PyThread_start_joinable_thread(thread_run, (void*) boot, &ident_obj->handle);
    } else {
        unsigned long ident = PyThread_start_new_thread(thread_run, (void*) boot);
        res = (ident == PYTHREAD_INVALID_THREAD_ID) ? -1 : 0;
        ident_obj->handle.legacy_ident = ident;
    }
    if (res < 0) {
        PyErr_SetString(ThreadError, "can't start new thread");
        goto error;
    }
    return ident_obj;

error:
    Py_XDECREF(ident_obj);
    PyThreadState_Clear(boot->tstate);
    thread_bootstate_free(boot, 1);
    return NULL;
```

`_PyThreadID_Type` would act like `PyLong_Type` even if subclassing it wouldn't be feasible (at least I think it isn't).  We would make it as close a drop-in replacement as is practical:

* define members/getsets to expose the raw handle as an int object in the various ways we care about (unsigned long, etc.)
* define `__int__`
* define comparison (and reverse comparison) operators so it compares correctly with `PyLongObject`, etc.
* anything else

Reply:
I've switched to a thread handle object now. It's slightly different from your proposal, but I think the underlying idea is the same.

Reply:
Ultimately, I think using a struct down at the `PyThread_*` level (and then having a `PyObject` that *wraps* it) solves the most problems, but the approach you've taken is definitely an improvement. 😄  Since the involved API are internal now: *If* we want to take things further, we can do so later.

Reply:
I'm not sure a struct would achieve much apart from making the APIs slightly more elegant perhaps. It's not like we're in C++ and can define methods and a destructor on the struct.